tensorflow/contrib/verbs/rdma.cc

```diff
@@ -66,16 +69,332 @@ string MessageTypeToString(RdmaMessageType rmt) {
+    gid->raw[12] | gid->raw[13] | gid->raw[14] | gid->raw[15]);
+}
+
+// Function read from sysfs file
+// Args:
+//   dir - directory
+//   file - file
+//   buff - buffer for the result
+//   size - buffer size
+// Returns:
+//   number of bytes were read or -1 if failed
+int read_sysfs_file(const char *dir, const char *file, char *buf, size_t size) {
+  char *path;
+  int fd;
+  int len;
+
+  if (asprintf(&path, "%s/%s", dir, file) < 0)
+    return -1;
+
+  fd = open(path, O_RDONLY);
+  if (fd < 0) {
+    free(path);
+    return -1;
+  }
+
+  len = read(fd, buf, size);
+
+  close(fd);
+  free(path);
+
+  if (len > 0 && buf[len - 1] == '\n')
+    buf[--len] = '\0';
+
+  return len;
+}
+
+// Function to check if GID index support RoCE V2
+// Args:
+//   context - device context
+//   port_num - port number
+//   index -  GID index
+// Returns:
+//   if GID supports RoCE V2 - true, otherwise - false.
+bool is_gid_type_roce_v2(ibv_context* context, uint8_t port_num, uint8_t index) {
+  char name[32];
+  char buff[41];
+
+  snprintf(name, sizeof(name), "ports/%d/gid_attrs/types/%d",  port_num, index);
+  if (read_sysfs_file(context->device->ibdev_path, name, buff, sizeof(buff)) <= 0) {
+      return false;
+  }
+  return !strcmp(buff, RoCE_V2);
+}
+
+// Function to set GID index.
+// If the port link is IB, no GID index should be selected.
+// If Ethernet but RDMA_GID_INDEX not set gid index that supports
+//   RoCE V2 will be chosen(fails if more then one IP is configured)
+// Args:
+//   context - device context
+//   port_num - port number
+// Returns:
+//   GID index to use
+uint8_t set_gid(uint8_t port_num, ibv_context* context) {
+  ibv_port_attr port_attr;
+  string gid_str;
+  int rc, i, gids_num = 0, v2_ip_num = 0;
+  union ibv_gid gid;
+  uint8_t gid_index = 0;
+
+  rc = ibv_query_port(context, port_num, &port_attr);
+  CHECK(!rc) << "Failed to query the port" << port_num;
+
+  for (i = 0; i < port_attr.gid_tbl_len; i++) {
+    rc = ibv_query_gid(context, port_num, i, &gid);
+    CHECK(!rc) << "Failed to query gid to port " << (int)port_num << " index " <<  i;
+    if (!is_gid_null(&gid)) {
+      gids_num++;
+      if (gid.raw[0] == 0 && gid.raw[1] == 0 && is_gid_type_roce_v2(context, port_num, i)) {
+        if (v2_ip_num == 0) {
+        //can be overwritten by RDMA_GID_INDEX later
+        gid_index = i;
+        }
+        v2_ip_num++;
+      }
+    }
+  }
+  switch(port_attr.link_layer){
+    case (IBV_LINK_LAYER_ETHERNET):
+      gid_str = get_env_var("RDMA_GID_INDEX");
+      if (!gid_str.empty()) {
+        gid_index = stoi(gid_str);
+        CHECK(gid_index < gids_num) << "RDMA_GID_INDEX should be less than GIDs amount" << gids_num;
+      }
+      else {
+        CHECK(v2_ip_num <= 1) << "More than one IP is available, please specify GID_INDEX";
+      }
+      break;
+    case (IBV_LINK_LAYER_INFINIBAND): //no need in GID index
+      break;
+    default:
+      CHECK(false) << "Unknown port link layer!";
```

Comment:
I think we should remove the default CHECK (failing if it's not the IB/ETH LINKs).
We may have something else in exp_verbs